File: Binary Tree/c++/tree_includes.cpp

#include<bits/stdc++.h>
using namespace std;

class Node
{
    public:
        string val;
        Node* left;
        Node* right;

        Node(string initialVal)
        {
            val=initialVal;
            left=nullptr;
            right=nullptr;
        }
};

std::vector<std::string> depthFirstValues(Node* root)
{
    if (root)
    {
        vector<string>re{root->val};
        vector<string>left=depthFirstValues(root->left);
        vector<string>right=depthFirstValues(root->right);
        re.insert(re.end(),left.begin(),left.end());
        re.insert(re.end(),right.begin(),right.end());

        return re;
    }

    return {};
}

bool treeIncludes(Node* root, std::string targetVal)
{
    if (root)
    {
        if (root->val==targetVal) return true;
        if (treeIncludes(root->left,targetVal)) return true;
        if (treeIncludes(root->right,targetVal)) return true;
    }

    return false;
}

int main()
{

}

//https://structy.net/problems/tree-includes

File: Binary Tree/python/tree_includes.py
class Node:
    def __init__(self, val):
        self.val=val
        self.left=None
        self.right=None

def tree_includes(root, target):
    if root:
        if root.val==target:
            return True
        
        if tree_includes(root.left,target): return True
        if tree_includes(root.right,target): return True

    return False

a = Node("a")
b = Node("b")
c = Node("c")
d = Node("d")
e = Node("e")
f = Node("f")

a.left = b
a.right = c
b.left = d
b.right = e
c.right = f

print(tree_includes(a, "e")) # -> True

#https://structy.net/problems/tree-includes
